apps/bfd-pipeline/bfd-pipeline-ccw-rif/src/main/java/gov/cms/bfd/pipeline/ccw/rif/extract/s3/DataSetManifestFactory.java

```diff
@@ -0,0 +1,106 @@
+          + "      </xs:sequence>"
+          + "      <xs:attribute name=\"timestamp\" type=\"xs:dateTime\" use=\"required\" />"
+          + "      <xs:attribute name=\"sequenceId\" type=\"xs:unsignedByte\" use=\"required\" />"
+          + "      <xs:attribute name=\"syntheticData\" type=\"xs:boolean\" use=\"optional\" />"
+          + "    </xs:complexType>"
+          + "  </xs:element>"
+          + "</xs:schema>";
+
+  /** The {@link DataSetManifestFactory} our factory instance (singleton). */
+  private static DataSetManifestFactory thisInstance;
+
+  /** private constructor...only we get to use it {@link DataSetManifestFactory}. */
+  private DataSetManifestFactory() {}
+
+  /**
+   * Constructs a new {@link DataSetManifestFactory} instance.
+   *
+   * @return the {@link DataSetManifestFactory}
+   */
+  public static DataSetManifestFactory newInstance() {
+    if (thisInstance == null) {
+      thisInstance = new DataSetManifestFactory();
+    }
+    return thisInstance;
+  }
+
+  /**
+   * Parses the input filename {@link String} and returns a new {@link DataSetManifest} instance.
+   *
+   * @param manifestFilename the {@link String} XML filename to read/parse
+   * @return the {@link DataSetManifest}
+   * @throws JAXBException (any errors encountered will be bubbled up).
+   */
+  public DataSetManifest parseManifest(String manifestFilename) throws JAXBException {
+    try {
+      InputStream manifestStream =
+          Thread.currentThread().getContextClassLoader().getResourceAsStream(manifestFilename);
+      return parseManifest(manifestStream);
+    } catch (Exception e) {
+      throw new JAXBException(e.getMessage(), e);
+    }
+  }
+
+  /**
+   * Parses the input byte stream {@link InputStream} and returns a new {@link DataSetManifest}
+   * instance. It validates the XML data stream {@link InputStream} to be both well-formed and
+   * valid.
+   *
+   * @param manifestStream the {@link InputStream} stream of XML to read/parse
+   * @return the {@link DataSetManifest}
+   * @throws JAXBException (any errors encountered will be bubbled up).
+   * @return the {@link DataSetManifest}
+   */
+  public DataSetManifest parseManifest(InputStream manifestStream) throws JAXBException {
+    DataSetManifest manifest = null;
+    try {
+      InputStream xsdStream = new ByteArrayInputStream(MANIFEST_XSD.getBytes());
+      SchemaFactory sf = SchemaFactory.newInstance(XMLConstants.W3C_XML_SCHEMA_NS_URI);
+      Schema schema = sf.newSchema(new StreamSource(xsdStream));
+      JAXBContext jaxbContext = JAXBContext.newInstance(DataSetManifest.class);
+      Unmarshaller jaxbUnmarshaller = jaxbContext.createUnmarshaller();
+      jaxbUnmarshaller.setSchema(schema);
+      manifest = (DataSetManifest) jaxbUnmarshaller.unmarshal(manifestStream);
+    } catch (SAXException e) {
+      throw new JAXBException(e.getMessage(), e);
+    } catch (Exception e) {
+      throw new JAXBException(e.getMessage(), e);
```

Comment:
I think we should avoid constructing JAXBExceptions ourselves -- it could be confusing for debugging purposes since most will assume that JAXBExceptions originate from JAXB.

Reply:
Depends....I didn't write the AC, but here is the pertinent part:

The XML binding library should handle the error cases above for us by **throwing a JAXBException which is already handled in the BFD pipeline.**

Reply:
It is not an AC, just an implementation note and it did not anticipate usage of XSD (which is a great idea btw) so we may need to tweak the caller to handle SAXException or wrap everything in some BFD flavor of exception.

Reply:
done.